jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py

```diff
@@ -284,6 +358,68 @@ def collapse_to_spectrum(self, add_data=True, **kwargs):
+                                               color="error", sender=self))
+            return
+
+        masks_weights = np.zeros_like(flux_cube.flux.value, dtype=np.float32)
+
+        # Center is reverse coordinates
+        center = (self.aperture.selected_spatial_region.center.y,
+                  self.aperture.selected_spatial_region.center.x)
+
+        # Loop through cube and create cone aperture at each wavelength. Then convert that to a
+        # weight array using the selected aperture method, and add it to a weight cube.
+        # TODO: Use flux_cube.spectral_axis.to_value(display_unit) when we have unit conversion.
+        radii = ((flux_cube.spectral_axis.value / self.reference_wavelength) *
+                 self.aperture.selected_spatial_region.radius)
+        im_shape = (flux_cube.shape[0], flux_cube.shape[1])
+        aper_method = self.aperture_method_selected.lower()
+        for index, radius in enumerate(radii):
+            aperture = CircularAperture(center, r=radius)
+            slice_mask = aperture.to_mask(method=aper_method).to_image(im_shape)
+            # Add slice mask to fractional pixel array
+            masks_weights[:, :, index] = slice_mask
+        return masks_weights
+
+    def cylindrical_aperture(self):
+        # Retrieve flux cube and create an array to represent the cone mask
+        flux_cube = self._app._jdaviz_helper._loaded_flux_cube.get_object(cls=Spectrum1D,
+                                                                          statistic=None)
+        # TODO: Replace with code for retrieving display_unit in cubeviz when it is available
+        display_unit = flux_cube.spectral_axis.unit
+        if display_unit.physical_type != 'length':
+            self.hub.broadcast(SnackbarMessage('Spectral axis unit physical type is '
+                                               f'{display_unit.physical_type}, must be length',
+                                               color="error", sender=self))
+            return
+
+        masks_weights = np.zeros_like(flux_cube.flux.value, dtype=np.float32)
+
+        # Center is reverse coordinates
+        center = (self.aperture.selected_spatial_region.center.y,
+                  self.aperture.selected_spatial_region.center.x)
+
+        # Create a slice_mask and set each slice of a numpy cube to that mask
+        # TODO: Use flux_cube.spectral_axis.to_value(display_unit) when we have unit conversion.
+        radius = self.aperture.selected_spatial_region.radius
+        im_shape = (flux_cube.shape[0], flux_cube.shape[1])
+        aper_method = self.aperture_method_selected.lower()
+        aperture = CircularAperture(center, r=radius)
+        slice_mask = aperture.to_mask(method=aper_method).to_image(im_shape)
+        for index in range(0, len(flux_cube.spectral_axis)):
+            # Add slice mask to fractional pixel array
+            masks_weights[:, :, index] = slice_mask
```

Comment:
Maybe you can avoid a Python loop like this? Do double check the broadcasting result first.

```suggestion
        masks_weights = np.repeat(slice_mask[:, :, np.newaxis], flux_cube.spectral_axis.size, axis=2)
```

ref: https://stackoverflow.com/questions/32171917/how-to-copy-a-2d-array-into-a-3rd-dimension-n-times

Reply:
I found something similar to replace this line.